docs/split-inputs.md

```diff
@@ -0,0 +1,140 @@
+
+If the data format is common (e.g. a widely used media format or network packet
+format) then it is highly desirable for a fuzz target to consume exactly this
+data format, and not some custom modification.
+This way it will be easier to procure a seed corpus for this fuzz target
+and to use the generated corpus to test/fuzz other targets.
+
+## Multiple options
+If the data format may be processed by a fuzz target in a small number of different ways,
+it is often the best approach to split the fuzz target into several ones,
+each processing the input in exactly one way.
+
+Make sure to cross-pollinate the corpora between these targets.
+OSS-Fuzz does that automatically.
+
+## Embedding / Comments
+When a fuzz target for a common data format requires some flags, options, or
+additional auxiliary sub-input(s), it is sometimes possible to embed the extra input
+inside a custom section or a comment of the main data format.
+
+Examples:
+* PNG allows custom "chunks", and so a fuzz target for a PNG decoder can
+  hide the flags used during PNG processing
+  in a separate PNG chunk, e.g. `fUZz`
+  ([example](https://github.com/google/oss-fuzz/blob/master/projects/libpng-proto/libpng_transforms_fuzzer.cc)).
+* When fuzzing C/C++/Java/JavaScript inputs one may hide a sub-input in a single-line `//` comment. TODO: example?
+
+
+## Hash
+When only one small fixed-size sub-input is required (such as flags / options),
+the fuzz target may compute a hash function on the full input and use it as the flag bits.
+This option is very easy to implement, but it's applicability is limited to
+relatively simple cases. The major problem is that a small local mutation of the input
+leads to a large change in the sub-input, which often makes fuzzing less
+efficient. Try this approach if the flags are individual bits and the input
+type allows some bit flips in the inputs (e.g. a plain text).
+
+TODO: example.
+
+# Custom Serialization Format
+
+If you **do not intend to share the corpus** with any other API or fuzz targets,
+then a custom serialization format might be a good option for a multi-input fuzz
+target.
+
+## First / Last Bytes
+
+When only one fixed-size sub-input is required (such as flags / options),
+it is possible to treat the first (or last) `K` bytes of the input as sub-input,
+and the rest of the bytes as the main input.
+
+Just remember to copy the main input into a separate heap buffer of `Size - K`
+bytes, so that buffer under/overflows on the main input are detected.
+
+TODO: example.
+
+## Magic separator
+
+Choose a 4-byte (or 8-byte) magic constant that will serve as a separator
+between the inputs.
+In the fuzz target, split the input using this separator. Use `memmem` to
+find the separator in the input -- `memmem` is known to be friendly to fuzzing
+engines, at least to libFuzzer.
+
+Example (see full code [here](https://github.com/llvm-mirror/compiler-rt/blob/master/test/fuzzer/MagicSeparatorTest.cpp)):
```

Comment:
bit: better use a fixed revision link, to make sure the file will be always accessible: https://github.com/llvm-mirror/compiler-rt/blob/6cd423889971c0d97801a9f3b9b5afb91ae9c137/test/fuzzer/MagicSeparatorTest.cpp